File: backend/src/main/java/com/zuber/organizeit/Model/Project.java
package com.zuber.organizeit.Model;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import lombok.*;

import javax.persistence.*;
import java.util.List;


@Entity
@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS)
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Project {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "project_id")
    @JsonProperty("project_id")
    private Long id;

    String name;

    @OneToOne(cascade = {CascadeType.PERSIST, CascadeType.REFRESH})
    @Builder.Default
    Task rootTask = Task.builder().description("What's poppin?").build();

    String description;

    @OneToMany(cascade = {CascadeType.ALL})
    List<Tag> tags;


}

File: frontend/src/Model/Project.js
import Task from './Task';

export class Project {
    constructor({project_id, title = "", description = "", root_task = Task.empty, tags = []}) {
        this.project_id = project_id;
        this.title = title;
        this.description = description;
        this.root_task = new Task(root_task);
        this.tags = tags;
        this['@class'] = "com.zuber.organizeit.Model.Project";
    }

    static ['@class'] = "com.zuber.organizeit.Model.Project"; // todo remove

    static empty() {
        return new Project({});
    }

    static recordToProject(projectRecord) {
        let project = Project.empty();
        if(projectRecord["@class"] === Project["@class"]) project = new Project({...projectRecord});
        return project;
    }

}